File: 09-backtracking/00039-Combination_Sum/00039-combination_sum.java
// 39: Combination Sum
// https://leetcode.com/problems/combination-sum/



import java.util.List;
import java.util.ArrayList;

class Solution {
    private void backtrack(int[] candidates, int target, int start, List<Integer> current, List<List<Integer>> result) {
        if (start >= candidates.length || target < 0) return;

        if (target == 0) {
            result.add(new ArrayList<Integer>(current));
            return;
        }

        current.add(candidates[start]);
        backtrack(candidates, target - candidates[start], start, current, result);
        current.remove(current.size()-1);
        backtrack(candidates, target, start+1, current, result);
    }

    // SOLUTION
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtrack(candidates, target, 0, current, result);
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] candidates = {2,3,6,7};
        int target = 7;

        // OUTPUT
        var result = o.combinationSum(candidates, target);
        System.out.println(result);
    }
}

File: 09-backtracking/00039-Combination_Sum/00039-combination_sum.cpp
// 39: Combination Sum
// https://leetcode.com/problems/combination-sum/



#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    void backtrack(vector<int>& candidates, int target, int start, vector<int>& current, vector<vector<int>>& result) {
        if (start >= candidates.size() || target < 0) return;

        if (target == 0) {
            result.push_back(current);
            return;
        }

        current.push_back(candidates[start]);
        backtrack(candidates, target - candidates[start], start, current, result);
        current.pop_back();
        backtrack(candidates, target, start+1, current, result);
    }

public:
    // SOLUTOIN
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> current;
        vector<vector<int>> result;
        backtrack(candidates, target, 0, current, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> candidates = {2,3,6,7};
    int target = 7;

    // OUTPUT
    auto result = o.combinationSum(candidates, target);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}
